internal/fs/dir_handle.go

```diff
@@ -156,78 +174,66 @@ func fixConflictingNames(entries []fuseutil.Dirent) (err error) {
 	return
 }
 
-// Read all entries for the directory, fix up conflicting names, and fill in
-// offset fields.
-//
-// LOCKS_REQUIRED(in)
-func readAllEntries(
-	ctx context.Context,
-	in inode.DirInode) (entries []fuseutil.Dirent, err error) {
-	// Read one batch
-	var batch []fuseutil.Dirent
-	batch, ContinuationToken, err = in.ReadEntries(ctx, ContinuationToken)
-	if err != nil {
-		err = fmt.Errorf("ReadEntries: %w", err)
-		return
-	}
-	// Accumulate.
-	entries = append(entries, batch...)
-
-	// Ensure that the entries are sorted, for use in fixConflictingNames
-	// below.
-	// TODO: Fix this after  asynchronous fetch is added.
-	sort.Sort(sortedDirents(entries))
-
-	// Fix name conflicts.
-	err = fixConflictingNames(entries)
-	if err != nil {
-		err = fmt.Errorf("fixConflictingNames: %w", err)
-		return
-	}
-
-	// Return a bogus inode ID for each entry, but not the root inode ID.
-	//
-	// NOTE(jacobsa): As far as I can tell this is harmless. Minting and
-	// returning a real inode ID is difficult because fuse does not count
-	// readdir as an operation that increases the inode ID's lookup count and
-	// we therefore don't get a forget for it later, but we would like to not
-	// have to remember every inode ID that we've ever minted for readdir.
-	//
-	// If it turns out this is not harmless, we'll need to switch to something
-	// like inode IDs based on (object name, generation) hashes. But then what
-	// about the birthday problem? And more importantly, what about our
-	// semantic of not minting a new inode ID when the generation changes due
-	// to a local action?
-	for i, _ := range entries {
-		entries[i].Inode = fuseops.RootInodeID + 1
-	}
-
-	return
-}
+// Fetch Dirent entries from GCSfuse.Will be used as a goroutine which is run asynchronously
+// to fetch data in the background while kernel requests are also served simultaneously.
+func (dh *DirHandle) FetchEntriesAsync(
+	rootInodeId int,
+	firstCall bool) {
+	ctx, cancel := context.WithCancel(context.TODO())
+	var err error
+	for ContinuationToken != "" || firstCall {
+		var entries []fuseutil.Dirent
+		dh.In.Lock()
+		entries, ContinuationToken, err = dh.In.ReadEntries(ctx, ContinuationToken)
+		dh.In.Unlock()
+		if err != nil {
+			err = fmt.Errorf("ReadEntries: %w", err)
+			rec.Lock()
+			rec.err = err
+			rec.Unlock()
+			rec.cond.Broadcast()
+			cancel()
+			runtime.Goexit()
```

Comment:
why do you need this?

Reply:
Incase of error conditions, the list call request should be stopped since error was during fetching. So , the function must exit . To prevent goroutine leak,I used the goexit function to kill the process fetching entries since this routine is independent of the goroutines serving kernel requests.

Reply:
what will happen if you just come out of forloop?

Reply:
Just coming out of loop after broadcasting should also work since goroutines are automatically terminated when they return.